### PR TITLE
Fix/SK-1127 | add session_id flag

### PR DESCRIPTION
**Pull request by KatHellg**
added session_id flag in model_cmd.py, round_cmd.py, statuses_cmd.py, and validation_cmd.py. Works well but might be a brute force solution so please lmk if there's a smarter way of doing it.

also added an id flag in each file to be able to search based on id (e.g., model_id in model_cmd.py)

### CODE REVIEW

**Review by niklastheman — CHANGES_REQUESTED** [review]
Cool, I think I would like to have the "id functionality" (list_clients for example) refactored in separate functions (not part of list). The implementation of this looks good! The "session_id functionality" can be implemented in an other way. The GET [entity]/ endpoints accepts url params for filtering. For example: GET /rounds?session_id=<session_id_value>

File: fedn/cli/client_cmd.py [hunk 1]
```diff
@@ -58,6 +57,38 @@ def list_clients(ctx, protocol: str, host: str, port: str, token: str = None, id
...
+
+
+@click.option("-p", "--protocol", required=False, default=CONTROLLER_DEFAULTS["protocol"], help="Communication protocol of controller (api)")
+@click.option("-H", "--host", required=False, default=CONTROLLER_DEFAULTS["host"], help="Hostname of controller (api)")
+@click.option("-P", "--port", required=False, default=CONTROLLER_DEFAULTS["port"], help="Port of controller (api)")
+@click.option("-t", "--token", required=False, help="Authentication token")
+@click.option("-id", "--id", required=False, help="Client ID")
+@client_cmd.command("get")
+@click.pass_context
+def get_client(ctx, protocol: str, host: str, port: str, token: str = None, id: str = None):
```
Comment (niklastheman) [inline fedn/cli/client_cmd.py:80]: Think id should be required...

File: fedn/cli/model_cmd.py [hunk 2]
```diff
@@ -17,24 +17,62 @@ def model_cmd(ctx):
...
         headers["X-Limit"] = n_max
 
+    _token = get_token(token)
+
+    if _token:
+        headers["Authorization"] = _token
+
+    if session_id:
+        url = f"{url}?session_id={session_id}"
+        headers["session_id"] = session_id
```
Comment (niklastheman) [inline fedn/cli/model_cmd.py:46]: This header should not be needed.

File: fedn/cli/package_cmd.py [hunk 3]
```diff
@@ -72,13 +103,10 @@ def list_packages(ctx, protocol: str, host: str, port: str, token: str = None, i
         headers["id"] = id
 
 
-    click.echo(f"\nListing packages: {url}\n")
+    click.echo(f"\nretrieving package: {url}\n")
```
Comment (niklastheman) [inline fedn/cli/package_cmd.py:106]: retrieving => Retrieving? :)

File: fedn/cli/client_cmd.py [hunk 4]
```diff
@@ -66,18 +97,14 @@ def list_clients(ctx, protocol: str, host: str, port: str, token: str = None, id
         headers["id"] = id
```
Comment (niklastheman) [inline fedn/cli/client_cmd.py:97]: This header should not be needed.

**Review by niklastheman — CHANGES_REQUESTED** [review]
Cool, good work! I have some comments, minor things...

File: fedn/cli/client_cmd.py [hunk 5]
```diff
@@ -65,12 +65,46 @@ def list_clients(ctx, protocol: str, host: str, port: str, token: str = None, n_
...
+
+    if _token:
+        headers["Authorization"] = _token
+
+    if id:
+        url = f"{url}{id}"
+
+
+    click.echo(f"\nRetrieving client: {url}\n")
+    click.echo(f"Headers: {headers}")
```
Comment (Wrede) [inline fedn/cli/client_cmd.py:99]: remove print of hearder

File: fedn/cli/client_cmd.py [hunk 6]
```diff
@@ -65,12 +65,46 @@ def list_clients(ctx, protocol: str, host: str, port: str, token: str = None, n_
 
     try:
         response = requests.get(url, headers=headers)
-        print_response(response, "clients")
+        print_response(response, "clients", None)
     except requests.exceptions.ConnectionError:
```
Comment (Wrede) [inline fedn/cli/client_cmd.py:69]: We should handle more than just ConnectionError, there are plenty of other error related to http requests. You don't have do do this know but make an other issue of it. For example create a utility function that handles all requests and checks the response for error:


import requests
from requests.exceptions import HTTPError, Timeout, ConnectionError, RequestException

def make_request(method, url, **kwargs):
    """
    A reusable function to handle HTTP(S) requests with error handling.

    Args:
        method (str): HTTP method (e.g., 'GET', 'POST').
        url (str): The URL for the request.
        **kwargs: Additional arguments passed to `requests.request()`.

    Returns:
        dict: A dictionary with the response data or an error message.
    """
    try:
        response = requests.request(method, url, **kwargs)
        response.raise_for_status()
        # Assuming JSON response, adjust if needed
        return {"success": True, "data": response.json()}
    except HTTPError as http_err:
        return {"success": False, "error": f"HTTP error occurred: {http_err}"}
    except Timeout:
        return {"success": False, "error": "The request timed out"}
    except ConnectionError:
        return {"success": False, "error": "Connection error occurred"}
    except RequestException as req_err:
        return {"success": False, "error": f"Request error: {req_err}"}
    except Exception as e:
        return {"success": False, "error": f"Unexpected error: {e}"}

File: fedn/cli/combiner_cmd.py [hunk 7]
```diff
@@ -88,11 +88,46 @@ def list_combiners(ctx, protocol: str, host: str, port: str, token: str = None,
...
+
+    if _token:
+        headers["Authorization"] = _token
+
+    if id:
+        url = f"{url}{id}"
+
+
+    click.echo(f"\nRetrieving combiner: {url}\n")
+    click.echo(f"Headers: {headers}")
```
Comment (Wrede) [inline fedn/cli/combiner_cmd.py:128]: remove

File: fedn/cli/model_cmd.py [hunk 8]
```diff
@@ -38,11 +41,49 @@ def list_models(ctx, protocol: str, host: str, port: str, token: str = None, n_m
...
+    _token = get_token(token)
+
+    if _token:
+        headers["Authorization"] = _token
+
+    if id:
+        url = f"{url}{id}"
+
+    click.echo(f"\nRetrieving model: {url}\n")
+    click.echo(f"Headers: {headers}")
```
Comment (Wrede) [inline fedn/cli/model_cmd.py:84]: remove

File: fedn/cli/shared.py [hunk 9]
```diff
@@ -64,26 +64,36 @@ def get_client_package_dir(path: str) -> str:
...
     """Prints the api response to the cli.
     :param response:
         type: array
         description: list of entities
     :param entity_name:
         type: string
         description: name of entity
+    :param so:
+        type: boolean
+        desriptions: single output format (y/n)
```
Comment (Wrede) [inline fedn/cli/shared.py:77]: remove (y/n)

File: fedn/cli/shared.py [hunk 10]
```diff
@@ -64,26 +64,36 @@ def get_client_package_dir(path: str) -> str:
...
-            click.echo("{")
-            for k, v in obj.items():
+        if so:
+            click.echo(f"Found {entity_name}")
+            click.echo("\n---------------------------------\n")
+            for k, v in json_data.items():
                 click.echo(f"\t{k}: {v}")
-            click.echo("}")
+        else:
+            count, result = json_data.values()
```
Comment (Wrede) [inline fedn/cli/shared.py:88]: we should consider using json pretty prints instead, however I would prefer not introducing a new dependency. Save for other issue

**Review by Wrede — CHANGES_REQUESTED** [review]
Good! Just remove all headers prints, I don't know why we have them, more as a debug print it feels. Had some comments that you can make new issues from. Finally, you should add unittests for cli cmds to make sure they work as expected.

**Review by Wrede — APPROVED** [review]

**Review by niklastheman — APPROVED** [review]

**Pull request merged**